### PR TITLE
Create output method to access attributes of linear regression model

### LINKED ISSUE #9: MachineLearning_Toolkit: Linear regression inputs and outputs
#### Description:

Add support of multiple inputs.
Expose regression variables from model.

### PR DESCRIPTION
**Pull request by BingWangUS**
### Issues addressed by this PR

Closes #9

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/MachineLearning_Toolkit/MachineLearning%239-LinearRegressionOutputs?csf=1&web=1&e=SDq2Pn

### Changelog

Create a method to output attributes, coefficients and intercept, from the linear regression model. 
Correct the shape check function in the code, for multiple dimension input X.

### CODE REVIEW

File: MachineLearning_Engine/Compute/LinearRegression.cs [hunk 1]
```diff
@@ -57,5 +60,21 @@ public static Tensor Infer(LinearRegression model, Tensor x)
         }
 
         /*************************************/
+
+        [Description("Projects the given input using a linear regression model")]
+        [Input("model", "The linear regressor model used for inference")]
+        [Output("")]
+        public static Output<List<double>, double> Coefs(LinearRegression model)
+        {
+            PyObject coefs = BH.Engine.MachineLearning.Compute.Invoke("LinearRegression.coefs", model);
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/LinearRegression.cs]: I believe you can get the attributes of the model directly.
```c#
double coef = model.SkLearnModel.GetAttr("coef_").As<double>();
double intercept = model.SkLearnModel.GetAttr("intercept_").As<double>();
return new Output<double, double> { Item1 = coef, Item2 = intercept };
```

File: MachineLearning_Engine/Compute/LinearRegression.cs [hunk 2]
```diff
@@ -57,5 +60,21 @@ public static Tensor Infer(LinearRegression model, Tensor x)
         }
 
         /*************************************/
+
+        [Description("Projects the given input using a linear regression model")]
+        [Input("model", "The linear regressor model used for inference")]
+        [Output("")]
+        public static Output<List<double>, double> Coefs(LinearRegression model)
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/LinearRegression.cs]: Maybe we can move the method to the `Query` class?
If we name it `Weights` or `Coefficients` in a file like `Coefficients.cs` this can also be more scalable in the future to include SVCs, SVRs, DNNs, KMeans and so on :)

File: MachineLearning_Engine/MachineLearning_Engine.csproj [hunk 3]
```diff
@@ -48,6 +48,10 @@
       <HintPath>..\..\Python_Toolkit\Build\Python_Engine.dll</HintPath>
       <Private>False</Private>
     </Reference>
+    <Reference Include="Reflection_Engine, Version=3.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (epignatelli) [inline MachineLearning_Engine/MachineLearning_Engine.csproj]: Can you remove `Version`, `Culture` and `processorArchitecture`, and set `Copy Local` to false?

File: MachineLearning_Engine/Compute/LinearRegression.py [hunk 4]
```diff
@@ -25,8 +25,8 @@
 
 
 def fit(x: np.ndarray, y: np.ndarray):
-	assert x.shape == y.shape, "Input data and target data have different shape {} and {}".format(
-								x.shape, y.shape)
+	assert x.shape[0] == y.shape[0], "Input data and target data have different shape {} and {}".format(
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/LinearRegression.py]: Uhm, not sure about this change. What was the issue with checking the entire shape? This way it only controls the number of samples, correct?

File: MachineLearning_Engine/Compute/LinearRegression.py [hunk 5]
```diff
@@ -41,6 +41,8 @@ def fit(x: np.ndarray, y: np.ndarray):
 	model.fit(x, y)
 	return model
 
+def coefs(model: LinearRegression):
```
Comment (epignatelli) [inline MachineLearning_Engine/Compute/LinearRegression.py]: If you get the attributes directly as in the comment above, we can also simplify by removing this, as it becomes redundant!

**Review by epignatelli — CHANGES_REQUESTED** [review]
This is amazing @BingWangUS ! 
Thanks very much for getting stuck in! I left some comments on the code that may simplify things a bit!

File: MachineLearning_Engine/Compute/LinearRegression.cs (hunk 1 above)
Comment (BingWangUS) [reply MachineLearning_Engine/Compute/LinearRegression.cs]: This is clearer. 
However, having the same issue that need a conversion between the variable given by .GetAttr and c# type.
If we print the coef we will get "[[ 2.78734161e-01 -2.13865448e-17]]" for a 2 dimension regression, and "[1.]" for the intercept. 
Instead of handling the conversion like I did, we should have a conversion method.

File: MachineLearning_Engine/Compute/LinearRegression.py (hunk 4 above)
Comment (BingWangUS) [reply MachineLearning_Engine/Compute/LinearRegression.py]: The original one will fail for 2 dimension tensors. When I plug 2 dimension tensor in, it is giving error message that x has shape (50,2) while y has shape (50,).

File: MachineLearning_Engine/Query/Coefficients.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+{
+    public static partial class Query
+    {
+        /*************************************/
+        /**** Public Fields              ****/
+        /*************************************/
+
+        [Description("Expose the variables for the given regression model.")]
+        [Input("model", "The linear regressor model used for inference")]
+        [Output("")]
```
Comment (FraserGreenroyd) [inline MachineLearning_Engine/Query/Coefficients.cs]: Output name and description is needed

File: MachineLearning_Engine/Query/Coefficients.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        public static Output<List<double>, double> Coefficients(LinearRegression model)
+        {
+            BH.Engine.Reflection.Compute.RecordNote(model.SkLearnModel.GetAttr("coef_").ToString());
+            BH.Engine.Reflection.Compute.RecordNote(model.SkLearnModel.GetAttr("intercept_").ToString());
```
Comment (FraserGreenroyd) [inline MachineLearning_Engine/Query/Coefficients.cs]: Why are we recording these notes?

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
See comments on code.

PR title also needs improvement - `coef` and `statics` make no sense to me so are unlikely to make sense to others reading the change log. PR title should always be human readable for the scraped change log.

File: MachineLearning_Engine/Query/Coefficients.cs (hunk 7 above)
Comment (BingWangUS) [reply MachineLearning_Engine/Query/Coefficients.cs]: This is a WIP PR (just added a label). This is checking the output for testing purpose. Will remove later.

File: MachineLearning_Engine/Compute/LinearRegression.cs (hunk 1 above)
Comment (epignatelli) [reply MachineLearning_Engine/Compute/LinearRegression.cs]: You can use `BH.Engine.Python.Convert.IFromPython` if you don't know the type that you will receive. Or, in alternative, you can return a `Tensor` out of it
`Tensor coeff = new Tensor(model.SkLearnModel.GetAttr("intercept_"));`

File: MachineLearning_Engine/Compute/LinearRegression.py (hunk 4 above)
Comment (epignatelli) [reply MachineLearning_Engine/Compute/LinearRegression.py]: Oh, sorry this is my bad, agreed! 
Maybe then `len(x) == len(y)` is a better choice?

File: MachineLearning_Engine/Compute/LinearRegression.cs (hunk 1 above)
Comment (BingWangUS) [reply MachineLearning_Engine/Compute/LinearRegression.cs]: Great. Tensor then to list will work!

**Comment by BingWangUS** [comment]
After taking Eduardo's suggestion in code, this method is ready for review now.

**Review by FraserGreenroyd — COMMENTED** [review]
I'm happy with what I asked for changes on, will leave @epignatelli to do final review and merge

**Review by epignatelli — APPROVED** [review]
Great exciting job, thanks @BingWangUS !

**Review by FraserGreenroyd — APPROVED** [review]
Sorry, forgot to dismiss my original review - my comment was an approval pending @epignatelli so am happy to give this a final thumbs up for merge now that he's happy :smile:
